text/nodejs/0013-install-bom.md

```diff
@@ -0,0 +1,116 @@
+Yarn Install buildpacks in order to provide entries for modules installed
+during the build process.
+
+## Motivation
+
+As a part of an initiative to provide a Bill of Materials (BOM) in the Node.JS
+Language Family, we have to support BOM generation in both buildpacks that
+directly install dependencies (Node Engine and Yarn), as well as in buildpacks
+that run an installion command that results in a set of installed dependencies
+(NPM Install and Yarn Install)
+
+For the directly installed dependencies in the Node Engine and Yarn buildpacks,
+we control the metadata around the dependencies and largely implemented BOM
+metadata retrieval on the dependency server side. The availability of this
+metadata made it straightforward to consume the metadata and generate the BOM.
+
+For the indirectly installed modules that are installed by the NPM Install and
+Yarn Install buildpacks, this is not the case because we delegate the
+installation process to the package managers. Due to the nature of these
+dependencies and how they are installed, the process of generating a Bill of
+Materials will require changes to the installation buildpacks themselves.
+
+This is an important piece of the Bill of Materials, because it gives a
+complete picture of what dependencies are inside of the final application
+image. Without this metadata, a major value-add of a Bill of Materials is lost
+and users may have to use third-party tools to generate the rest of the BOM,
+which is a less than ideal user experience.
+
+
+## Implementation
+
+### Tool details
+
+The Bill of Materials that will result from the NPM/Yarn Install buildpacks
+should be an accurate accounting of all of the modules that get installed into
+the `node_modules` directory.
+
+There already exists a tool called [CycloneDX Node.js
+Module](https://github.com/CycloneDX/cyclonedx-node-module) which is an
+official CycloneDX supported-tool that creates valid CycloneDX BOMs for Node.JS
+applications. CycloneDX is a BOM format supported by the OWASP Foundation, and
+is a serious contender (at the time of this RFC creation) as a format we will
+eventually support.
+
+The CycloneDX Node.js Module tool can easily be installed via `npm install` and
+then run from a directory containing `node_modules`. The output BOM contains
+the following fields for each module:
+* name,
+* version,
+* description, 
+* hash, 
+* source URI, 
+* license ID, 
+* and package URL
+
+All of these are fields that we aim to support per [the Paketo BOM
+RFC](https://github.com/sophiewigmore/rfcs/blob/8b1e8c9ed6201313f47c5897223cbffb265e96ed/text/0028-bill-of-materials.md)
+
+### Buildpack Additions
+
+Using the CycloneDX tool to generate the BOM will involve some changes to the
+NPM Install and Yarn Install Buildpacks.
+
+* After the `npm install` or `yarn install` command is run during the build
+  phase, we should run `npm install -g @cyclonedx/bom` to install the tool on
```

Comment:
NPM Install/Yarn Install buildpacks are designed such that they can be used offline in Internet-less environments. This installation of cyclonedx/bom doesn't seem to work in offline mode. Can this be packaged as a dependency? Or even, do you think this merits a buildpack of its own or be in the build image? Regardless, I think we should maintain install buildpack's ability to work fully in offline mode.

Reply:
@arjun024 that's a good point. I'm more inclined to have this run as a part of this buildpack since the goal is just to collect insights around the install process, and it doesn't really have a role in adding something to the final app image (besides metadata). 

I can look into packaging this as a dependency and update you here.

Reply:
I think @ForestEckhardt and I have two options we are considering after looking into this:

A. Extract the tool as a tarball with its dependencies, and make it available as a standalone dep-server dependency. Install it with it’s own buildpack in both an online and offline scenario. Run BOM creation using the tool in its own buildpack after NPM Install/Yarn Install/ or before Node Start in the no package manager scenario. 

B. Extract the tool as a tarball with its dependencies, and then run BOM tool installation and BOM creation together in one buildpack after NPM Install/Yarn Install/ before Node Start in the no package manager scenario.

We are leaning towards having this process be a part of its own buildpack(s) for a couple reasons. First, it keeps buildpack modularity intact. While it seems intuitive to run BOM creation alongside the NPM Install and Yarn Install buildpacks, the issue is that we need to support the Node Start case, and it seems unintuitive to create the BOM in this buildpack, since it's a start-command buildpack.
 @arjun024 do you have any thoughts?
The other reason we are leaning to one of these two set ups is because it feels slightly sneaky to slip the dependency into the image within another buildpack. We could add it as a part of the node-engine dependency or install it in the node-engine buildpack, but this doesn't seem that transparent especially since the inclusion of this dependency opens the door for additional CVEs.

Reply:
We are ideally going with option B here. We will host the CycloneDX BOM tool as a dependency in the dep-server, and then introduce a buildpack that runs before start commands to install the tool and run BOM generation together.

Reply:
SGTM